### PR TITLE
Technical/add support for update on map

### PR DESCRIPTION
**Pull request by hasmikharutyunyan**

### CODE REVIEW

File: HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs [hunk 1]
```diff
@@ -88,19 +89,27 @@ private async Task MapCountries(Suppliers supplier, CancellationToken cancellati
                     Modified = utcDate
                 };
 
-                // Maybe after testing we will change to get data from db
-                var cached = await _countriesCache.Get(code);
-                if (cached != default)
+                var existing = notSupplierCountries.FirstOrDefault(c => c.Code == code);
+                var existingOfSupplier = supplierCountries.FirstOrDefault(c => c.Code == code);
```
Comment (kirillta) [inline HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs:93]: `existing` and `existingOfSupplier` names are messy

File: HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs [hunk 2]
```diff
@@ -150,133 +211,251 @@ private async Task MapLocalities(Suppliers supplier, CancellationToken cancellat
...
                     var defaultCountryName = locality.CountryNames.GetValueOrDefault(DefaultLanguageCode);
-                    // TODO: review and remove 
-                    var countryCode =
-                        _locationNameNormalizer.GetNormalizedCountryCode(defaultCountryName, locality.CountryCode);
                     var defaultLocalityName = locality.LocalityNames.GetValueOrDefault(DefaultLanguageCode);
                     var normalizedLocalityName =
                         _locationNameNormalizer.GetNormalizedLocalityName(defaultCountryName, defaultLocalityName);
 
-                    var cached = await _localitiesCache.Get(countryCode, normalizedLocalityName);
+                    var existing = notSupplierLocalities.FirstOrDefault(l => l.Names.En == normalizedLocalityName);
```
Comment (kirillta) [inline HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs:218]: Same here

File: HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs [hunk 3]
```diff
@@ -150,133 +211,251 @@ private async Task MapLocalities(Suppliers supplier, CancellationToken cancellat
...
+                            new Dictionary<Suppliers, string>(existing.SupplierLocalityCodes);
                         dbLocality.SupplierLocalityCodes.TryAdd(supplier, locality.LocalityCode);
-                        existingLocalities.Add(dbLocality);
+                        if (existingOfSupplier != default)
+                        {
+                            changedLocalityPairs.Add(existingOfSupplier.Id, existing.Id);
+
+                            foreach (var sup in existingOfSupplier.SupplierLocalityCodes)
+                                dbLocality.SupplierLocalityCodes.TryAdd(sup.Key, sup.Value);
+                            existingOfSupplier.IsActive = false;
```
Comment (kirillta) [inline HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs:242]: L blank line is missing

File: HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs [hunk 4]
```diff
@@ -150,133 +211,251 @@ private async Task MapLocalities(Suppliers supplier, CancellationToken cancellat
...
+                    var defaultLocalityName = zone.LocalityNames.GetValueOrDefault(DefaultLanguageCode);
+                    var normalizedLocalityName =
+                        _locationNameNormalizer.GetNormalizedLocalityName(defaultCountryName, defaultLocalityName);
+                    var defaultLocalityZone = zone.LocalityZoneNames.GetValueOrDefault(DefaultLanguageCode);
+                    var normalizedLocalityZone = defaultLocalityZone.ToNormalizedName();
 
-            var existingDbZones = await _context.LocalityZones
-                .Where(lz => normalizedKeyNames.Contains(lz.Names.En + lz.LocalityId))
-                .ToListAsync();
+                    var existing = notSupplierLocalityZones.FirstOrDefault(lz
```
Comment (kirillta) [inline HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs:372]: And here

File: HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs [hunk 5]
```diff
@@ -150,133 +211,251 @@ private async Task MapLocalities(Suppliers supplier, CancellationToken cancellat
...
+                            MultiLanguageHelpers.Merge(dbLocalityZone.Names, existing.LocalityZone.Names);
+                        dbLocalityZone.SupplierLocalityZoneCodes =
+                            new Dictionary<Suppliers, string>(existing.LocalityZone.SupplierLocalityZoneCodes);
+                        dbLocalityZone.SupplierLocalityZoneCodes.TryAdd(supplier, zone.LocalityZoneCode);
+                        if (existingOfSupplier != default)
+                        {
+                            changedLocalityZonesPairs.Add(existingOfSupplier.LocalityZone.Id, existing.LocalityZone.Id);
+                            foreach (var sup in existingOfSupplier.LocalityZone.SupplierLocalityZoneCodes)
+                                dbLocalityZone.SupplierLocalityZoneCodes.TryAdd(sup.Key, sup.Value);
+                            existingOfSupplier.LocalityZone.IsActive = false;
```
Comment (kirillta) [inline HappyTravel.StaticDataMapper.Api/Services/Workers/LocationMapper.cs:400]: Formatting

File: HappyTravel.StaticDataMapper.Data/NakijinContext.cs [hunk 6]
```diff
@@ -94,8 +94,8 @@ protected override void OnModelCreating(ModelBuilder builder)
             builder.Entity<AccommodationUncertainMatches>(m =>
             {
                 m.HasKey(p => p.Id);
-                m.Property(p => p.ExistingHtId).IsRequired();
-                m.Property(p => p.NewHtId).IsRequired();
+                m.Property(p => p.FirstHtId).IsRequired();
+                m.Property(p => p.SecondHtId).IsRequired();
```
Comment (kirillta) [inline HappyTravel.StaticDataMapper.Data/NakijinContext.cs:98]: Could they be `PrimaryId` and `SecondaryId`?

Comment (hasmikharutyunyan) [reply HappyTravel.StaticDataMapper.Data/NakijinContext.cs:98]: I think not, they are equal by meaning.

File: HappyTravel.StaticDataMapper.Api/HappyTravel.StaticDataMapper.Api.csproj [hunk 7]
```diff
@@ -24,7 +24,7 @@
         <PackageReference Include="HappyTravel.MultiLanguage" Version="1.0.2" />
         <PackageReference Include="HappyTravel.StdOutLogger" Version="1.0.15" />
         <PackageReference Include="HappyTravel.VaultClient" Version="1.0.5" />
-        <PackageReference Include="LocationNameNormalizer" Version="1.6.2" />
+        <PackageReference Include="LocationNameNormalizer" Version="1.6.4" />
```
Comment (mingazhev) [inline HappyTravel.StaticDataMapper.Api/HappyTravel.StaticDataMapper.Api.csproj:27]: Why not the latest?

File: HappyTravel.StaticDataMapper.Api/Infrastructure/StringComparisonHelper.cs [hunk 8]
```diff
@@ -11,17 +11,18 @@ public static float GetEqualityCoefficient(string first, string second, List<str
             if (string.IsNullOrWhiteSpace(first) || string.IsNullOrWhiteSpace(second))
                 return 0;
 
-            return GetSorensenDiceCoefficient(first.ToSequence(wordsToIgnore), second.ToSequence(wordsToIgnore));
+            return GetJaccardIndex(first.ToSequence(wordsToIgnore), second.ToSequence(wordsToIgnore));
         }
 
-        private static float GetSorensenDiceCoefficient(string[] firstSequence, string[] secondSequence)
+        private static float GetJaccardIndex(string[] firstSequence, string[] secondSequence)
```
Comment (mingazhev) [inline HappyTravel.StaticDataMapper.Api/Infrastructure/StringComparisonHelper.cs:17]: Could you add a comment with a link to the algorythm explanation?

File: HappyTravel.StaticDataMapper.Api/Services/Workers/AccommodationMapper.cs [hunk 9]
```diff
@@ -135,74 +127,105 @@ public async Task MapAccommodations(Suppliers supplier, CancellationToken cancel
...
+                _context.Entry(accommodationToUpdate).Property(p => p.IsCalculated).IsModified = true;
+                _context.Entry(accommodationToUpdate).Property(p => p.Modified).IsModified = true;
+
+                if (countryAccommodationsBySupplier.TryGetValue(accommodation.SupplierCode,
+                    out var existingAccommodation))
                 {
-                    dbAccommodation.SupplierAccommodationCodes.Add(supplier, accommodation.SupplierCode);
-                    dbAccommodation.IsCalculated = false;
-                    accommodationsToUpdate.Add(dbAccommodation);
+                    var accommodationToDeactivate = new RichAccommodationDetails
```
Comment (mingazhev) [inline HappyTravel.StaticDataMapper.Api/Services/Workers/AccommodationMapper.cs:180]: Could you split this to separate methods? It is hard for me to understand what's happenning

**Review by mingazhev — APPROVED** [review]

File: HappyTravel.StaticDataMapper.Api/Services/Workers/AccommodationMapper.cs (hunk 9 above)
Comment (hasmikharutyunyan) [reply HappyTravel.StaticDataMapper.Api/Services/Workers/AccommodationMapper.cs:180]: they are separated to local methods. You mean separate not local methods?

File: HappyTravel.StaticDataMapper.Api/Infrastructure/StringComparisonHelper.cs (hunk 8 above)
Comment (hasmikharutyunyan) [reply HappyTravel.StaticDataMapper.Api/Infrastructure/StringComparisonHelper.cs:17]: This is not the final verison, and I have not final and clear explanation why I choose this one.

File: HappyTravel.StaticDataMapper.Api/HappyTravel.StaticDataMapper.Api.csproj (hunk 7 above)
Comment (hasmikharutyunyan) [reply HappyTravel.StaticDataMapper.Api/HappyTravel.StaticDataMapper.Api.csproj:27]: It is the latest version of `LocationNameNormalizer`

**Pull request merged**